src/org/rust/lang/core/resolve/util/RustResolveUtil.kt

```diff
@@ -10,22 +13,49 @@ public object RustResolveUtil {
         while (current != null) {
             when (current) {
                 is RustResolveScope -> return current
-                else -> current = current.parent
+                else                -> current = current.parent
             }
         }
 
         return null
     }
 
-    fun getGlobalResolveScopeFor(elem: PsiElement): RustResolveScope {
-        var current = getResolveScopeFor(elem)
-        while (true)
-        {
-            val p = current?.let { getResolveScopeFor(it) } ?: break
-            current = p;
+    fun getCrateRootFor(elem: PsiElement): RustResolveScope? {
+        val mod = getSelfModFor(elem) ?: return null
+
+        val superMod = getSuperModFor(mod)
+        return if (superMod == null) mod else getCrateRootFor(superMod)
+    }
+
+    fun getSuperModFor(elem: RustModItem): RustModItem? {
```

Comment:
And, BTW, this one look suspiciously under-estimated in its complexity: traversing all the files to get all the `mod`-decls bringing up fully-fledged PSI isn't free-of-charge operation.